### PR TITLE
Alternative implementation without payoff function

### PR DESCRIPTION
**Pull request by fleupold**
Alternative implmentation to #1753 that doesn't use the extra payoff function (not sure if correct, since tests are failing, but I'm also not sure the test cases are correct)

### CODE REVIEW

File: crates/solver/src/settlement_rater.rs [hunk 1]
```diff
@@ -311,136 +311,203 @@ impl ScoreCalculator {
...
+            let payoff_expectation =
+                probability_success.clone() * payoff_success + probability_fail * payoff_fail;
+            tracing::trace!(?payoff_expectation, "Payoff expectation");
+            payoff_expectation
+        }
     }
 
     #[test]
     fn compute_score_with_success_probability_test() {
+        let cap = BigRational::from_float(1e17).unwrap();
```
Comment (fhenneke) [inline crates/solver/src/settlement_rater.rs:428]: The current cap is `1e16`.

File: crates/solver/src/settlement_rater.rs [hunk 2]
```diff
@@ -311,136 +311,203 @@ impl ScoreCalculator {
...
+    //
+    // => p_fail * std::cmp::min(bid, cap) = p_success * std::cmp::min(obj - bid,
+    // cap)
+    //
+    // We now look at the four cases that can arise from
+    // choosing the left or right argument in the `min` function on both
+    // sides of the equation:
+    if bid <= cap && objective.clone() - bid.clone() <= cap {
+        //Cap doesn't matter, equation reduces to the original one. We are done
+    } else if bid <= cap && objective.clone() - bid.clone() > cap {
```
Comment (fhenneke) [inline crates/solver/src/settlement_rater.rs:368]: The argument is not so simple, see [Haris' comment](https://github.com/cowprotocol/services/pull/1730#issue-1835206186) for correct case:
```suggestion
    } else if objective.clone() - bid.clone() > cap && probability_success < 0.5 {
```
This includes all the cases of you statement but also more cases, e.g. where `bid > cap && objective.clone() - bid.clone() > cap`.
One of the current unit tests should cover this.

File: crates/solver/src/settlement_rater.rs [hunk 3]
```diff
@@ -311,136 +311,203 @@ impl ScoreCalculator {
...
+        // In the success case our reward is capped, thus we need to solve for
+        // p_fail * bid = p_success * cap
+        bid = (probability_success * cap) / probability_fail;
+    } else if bid > cap && objective.clone() - bid.clone() <= cap {
```
Comment (fhenneke) [inline crates/solver/src/settlement_rater.rs:372]:
```suggestion
    } else if bid > cap && probabilty_success >= 0.5 {
```

File: crates/solver/src/settlement_rater.rs [hunk 4]
```diff
@@ -311,136 +311,203 @@ impl ScoreCalculator {
...
+        // In the failure case our penalty is capped, thus we need to solve
+        // p_fail * cap = p_success * (obj - bid)
+        bid = objective.clone() - (probability_fail * cap) / probability_success;
+    } else {
+        // The cap decides our payment in both cases, thus `bid`
+        // disappears from the equation and cannot be solved for.
+        // p_fail * cap = p_success * cap
+        // It can be proven by contradiction that this case can never happen.
+        bail!("Unreachable case for computing optimal bid");
```
Comment (fhenneke) [inline crates/solver/src/settlement_rater.rs:381]: This case does happen, see the unit tests for parameters.

**Review by fhenneke — CHANGES_REQUESTED** [review]
I am not opposed to ignore costs in case of failure (`cost_fail`) and then essentially follow the argument of Haris in #1730. This makes the logic a bit simpler.

One issue with the reasoning in this PR that it is not always enough to look at the bid without a cap and if the resulting payment would get capped. What matters is if the payment for the score you submit would get capped. In the second and third branch this works out automatically (as in: if you end up in this branch, then the bid as defined in the branch also satisfies the inequalities for ending up in the branch). For the fourth case it is not so simple and you can end up with a bid as in branch two or three, depending on the revert risk. The fourth branch can happen in practice (and is tested in unit tests).

File: crates/solver/src/settlement_rater.rs (hunk 4 above)
Comment (fleupold) [reply crates/solver/src/settlement_rater.rs:381]: Which unit tests are you referring to? I commented on #1753 that I wasn't able to fully understand where those are coming from (3 of them seem to test case 1)

**Comment by nlordell** [comment]
Closing as #1753 was merged.

**Pull request closed without merging**